### PR TITLE
schema: render schema paths from _CustomSafeLoaderWithMarks

### PR DESCRIPTION
**Pull request by blackboxsw**
## Proposed Commit Message
```
Use a custom yaml.SafeLoader to track SchemaMarks to define
start and end row/column values of each object within the
original cloud-config content.

Fixes KeyError from cloud-init devel schema --annotate when
encountering invalid nested list values under a dict key.

Drop _CustomSafeLoaderWithMarks replaces the need for
_schemapath_for_cloudconfig logic.

Fixes: SC-929
```

## Test Steps
```bash
$ cat <baduser.txt  >>EOF
#cloud-config
users:
- default                                                                       
- name: ok
  sudo: false
- name: user2
  sudo: 2
EOF
 PYTHONPATH=. python3 -m cloudinit.cmd.main devel schema -c baduser.txt --annotate

# expect valid `E1` failure annotations on the correct line of the original content
```

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

**Review by holmanb — COMMENTED** [review]
@blackboxsw Thanks for the PR.

I'm currently catching up on the context/purpose of our custom yaml classes for this review. Sorry question you on code you didn't write, but while doing some archeology to grok this change I noticed something odd in the commit history that I don't understand and I'm curious could help me understand.

I see that the same commit (cf30836645473c62599e838ab48b2d31677fa584) that added the safe dumper class also changed `yaml.safe_dump` to `yaml.dump` and elected to default to `yaml.dumper.Dumper` rather than `yaml.dumper.SafeDumper` as the default. This change doesn't really make sense to me in the context of the commit message. My understanding of the yaml api is superficial at present, so perhaps I misunderstand the usage. Would you happen to know offhand why this was changed from the "safe" api?

Aside: `safeyaml.py` seemed a bit light on comments / references in the code prior to your PR. Could we grab add some context from cf30836645473c62599e838ab48b2d31677fa584 and a7a9de1a079a70f5c8290ab5158661d3a33e5552 describing how this alters default yaml parsing while we are touching this code?

No objections so far, but I have more to review.

**Comment by blackboxsw** [comment]
> @blackboxsw Thanks for the PR.
> 
> I'm currently catching up on the context/purpose of our custom yaml classes for this review. Sorry question you on code you didn't write, but while doing some archeology to grok this change I noticed something odd in the commit history that I don't understand and I'm curious could help me understand.
> 
> I see that the same commit ([cf30836](https://github.com/canonical/cloud-init/commit/cf30836645473c62599e838ab48b2d31677fa584)) that added the safe dumper class also changed `yaml.safe_dump` to `yaml.dump` and elected to default to `yaml.dumper.Dumper` rather than `yaml.dumper.SafeDumper` as the default. This change doesn't really make sense to me in the context of the commit message. My understanding of the yaml api is superficial at present, so perhaps I misunderstand the usage. Would you happen to know offhand why this was changed from the "safe" api?
> 
> Aside: `safeyaml.py` seemed a bit light on comments / references in the code prior to your PR. Could we grab add some context from [cf30836](https://github.com/canonical/cloud-init/commit/cf30836645473c62599e838ab48b2d31677fa584) and [a7a9de1](https://github.com/canonical/cloud-init/commit/a7a9de1a079a70f5c8290ab5158661d3a33e5552) describing how this alters default yaml parsing while we are touching this code?
> 
> No objections so far, but I have more to review.

Thanks for the digging on this. I tried going through original PRs and even LP merge requests on this original content to refresh myself on the intent of those first branches. I agree with you there is something unique about the transition from yaml.safe_dump to a default of a yaml.Dumper which may have actually slipped through the cracks that conflicted with the original intent. I want to dig a bit more on this tomorrow as I'm inclined to move to the `yaml.dumper.SafeDumper` class for the default noalias=False behavior but I want to confirm use cases we know of  here w/ MAAS and any call-sites we have grown in cloudinit to make sure we aren't inadvertently breaking a behavior that has grown since the commit that moved us from safe_dump -> dump with `dumper=yaml.dumper.Dumper`.

**Comment by blackboxsw** [comment]
> > @blackboxsw Thanks for the PR.
> > I'm currently catching up on the context/purpose of our custom yaml classes for this review. Sorry question you on code you didn't write, but while doing some archeology to grok this change I noticed something odd in the commit history that I don't understand and I'm curious could help me understand.
> > I see that the same commit ([cf30836](https://github.com/canonical/cloud-init/commit/cf30836645473c62599e838ab48b2d31677fa584)) that added the safe dumper class also changed `yaml.safe_dump` to `yaml.dump` and elected to default to `yaml.dumper.Dumper` rather than `yaml.dumper.SafeDumper` as the default. This change doesn't really make sense to me in the context of the commit message. My understanding of the yaml api is superficial at present, so perhaps I misunderstand the usage. Would you happen to know offhand why this was changed from the "safe" api?
> > Aside: `safeyaml.py` seemed a bit light on comments / references in the code prior to your PR. Could we grab add some context from [cf30836](https://github.com/canonical/cloud-init/commit/cf30836645473c62599e838ab48b2d31677fa584) and [a7a9de1](https://github.com/canonical/cloud-init/commit/a7a9de1a079a70f5c8290ab5158661d3a33e5552) describing how this alters default yaml parsing while we are touching this code?
> > No objections so far, but I have more to review.
> 
> Thanks for the digging on this. I tried going through original PRs and even LP merge requests on this original content to refresh myself on the intent of those first branches. I agree with you there is something unique about the transition from yaml.safe_dump to a default of a yaml.Dumper which may have actually slipped through the cracks that conflicted with the original intent. I want to dig a bit more on this tomorrow as I'm inclined to move to the `yaml.dumper.SafeDumper` class for the default noalias=False behavior but I want to confirm use cases we know of here w/ MAAS and any call-sites we have grown in cloudinit to make sure we aren't inadvertently breaking a behavior that has grown since the commit that moved us from safe_dump -> dump with `dumper=yaml.dumper.Dumper`.

@holmanb, confirmed, that original changeset in 0836645473c62599e838ab48b2d31677fa584 inadvertently adapted the fallback option to move from `yaml.safe_dump` to yaml.dumper.Dumper as you suggested. There is nothing in callsites that would be negatively impacted by using the SafeDumper, so I'll shift this back to align more with the original implementation as I believe that was an oversight.

File: cloudinit/safeyaml.py [hunk 1]
```diff
@@ -4,14 +4,87 @@
...
+            )
+            if line_num not in self.schemamarks_by_line:
+                self.schemamarks_by_line[line_num] = [marks]
+            else:
+                self.schemamarks_by_line[line_num].append(marks)
+        return sequence
+
+    def get_single_data(self):
+        data = super().get_single_data()
+        if isinstance(data, dict):  # valid cloud-config scheam is a dict
```
Comment (holmanb) [inline cloudinit/safeyaml.py:80]: `"scheam" should be schema

**Review by holmanb — APPROVED** [review]
Reviewing this required some digging into how the yaml object internals work, but I don't see any issues or concerns with it.

The SafeLoader subclass just squirrels away mark start/end data in `schemamarks_by_line` for later use as it is parsing the yaml. I was previously going to ask for some extra comments around this code since the purpose isn't immediately obvious, but the yaml module code is pretty well documented, and future behavioral modifications to that code will require reading the yaml anyways so there really isn't much to add besides what you already have.

I ran a couple of speed and memory benchmarks on it, since parsing can potentially be computationally intensive, but the results showed insignificant difference from the `yaml.SafeLoader` class, so I'm not concerned at all about this.

Thanks for switching these to the "Safe" versions too.

Thanks for this!

**Pull request merged**